Workspace with Kobuki for Simulation and Real Robot Control

YuZhong-Chen opened the pull request:
Closes: #2 #14

j3soon commented on kobuki_ws/src/kobuki_navigation/config/ekf.yaml:
I'm not familiar with the configs in this `ekf.yaml` file for `robot_localization`, could you describe why these changes are necessary and potentially provide some references?

Another minor comment is the space style in `odom0_config` and `imu0_config` could be improved.

YuZhong-Chen replied:
The current style is:
```
odom0_config: [false, false, false,
               false, false, false,
                true,  true, false,
               false, false,  true,
               false, false, false]
```
Are you referring to a style similar to the one shown below?
```
odom0_config: [false, false, false,
               false, false, false,
               true,  true,  false,
               false, false, true,
               false, false, false]
```
or?
```
odom0_config: [false, false, false,
               false, false, false,
               true, true, false,
               false, false, true,
               false, false, false]
```

j3soon replied:
I believe the following style is more commonly used:

```
odom0_config: [false, false, false,
               false, false, false,
               true,  true,  false,
               false, false, true,
               false, false, false]
```

j3soon commented on kobuki_ws/src/kobuki_navigation/config/nav2_params.yaml:
Similarly, I'm not familiar with the configs in this `nav2_params.yaml` file for `nav2 costmap`, could you please also describe why these changes are necessary and point me to some references?

j3soon commented on kobuki_ws/src/kobuki_navigation/map/turtlebot3_stage_1.yaml:
Could you provide more details on how the PGM & YAML file are generated? Specifically, the commands and steps to generate these files.

YuZhong-Chen replied:
I believe the [nav2 tutorial](https://navigation.ros.org/tutorials/docs/navigation2_with_slam.html) provides sufficient details on how to generate the map file. 
The commands may vary for other users depending on how they install TurtleBot3.

For a standard installation, the commands in the tutorial should be sufficient.
```
# In terminal 1
ros2 launch turtlebot3_bringup robot.launch.py

# In terminal 2
ros2 launch nav2_bringup navigation_launch.py

# In terminal 3
ros2 launch slam_toolbox online_async_launch.py

# In terminal 4 ( Run this command after you have scanned the entire map and wish to save it. )
ros2 run nav2_map_server map_saver_cli -f ~/map
```

j3soon commented on kobuki_ws/src/kobuki_rviz/rviz/navigation.rviz:
Similar to previous comments, would prefer more readable RVIZ diffs. (Maybe separate copy & modify?)

j3soon reviewed (changes requested):
Thank you for this PR. I learned a lot while reviewing.

Aside from the comments, I have several questions arising from my unfamiliarity with the use of certain components.

<hr>

I will test this on a real hardware during the next review.

I've tested the simulation code, and it works mostly as expected. I did encounter some minor RVIZ issues, where the costmap seems to lag as the kobuki model rotates. I wonder if this issue can be resolved.

YuZhong-Chen commented on kobuki_ws/src/kobuki_rviz/launch/open_rviz_launch.py:
```diff
@@ -0,0 +1,36 @@
+    ld = LaunchDescription(ARGUMENTS)
+    ld.add_action(node_rviz)
+
+    return ld
```
The coding convention I followed comes from the [launch file of Husky](https://github.com/husky/husky/blob/729f8aa45ccd86fa33a05e07ef698c52c451cd9c/husky_gazebo/launch/gazebo.launch.py). Compared to the coding convention used in [TurtleBot3](https://github.com/ROBOTIS-GIT/turtlebot3/blob/a7dd05ae176f3f3778b0a36f7065dc9655b050e3/turtlebot3_navigation2/launch/navigation2.launch.py) or the [ROS2 tutorial](https://docs.ros.org/en/humble/Tutorials/Intermediate/Launch/Using-ROS2-Launch-For-Large-Projects.html), I prefer to use the current one. It is easier to scale because it is simpler to debug.

For example, the use of the pre-defined `ARGUMENTS` can help me quickly identify which parameters can be adjusted, and the use of `ld = LaunchDescription` allows me to easily enable or disable a specific node by modified only one line of code.

YuZhong-Chen replied:
However, I agree that using the coding convention from the official tutorial may make the code more readable and simpler sometimes, especially if there are a large number of files. 

Both coding conventions have their reasons. I'm not sure which one I should use here.
What are your thoughts on this?

j3soon replied:
Thanks for the references. I also prefer the current style. To summarize:

- The `ARGUMENTS` allow quickly view the adjustable parameters.
  - In contrast to [ROS2 tutorial](https://docs.ros.org/en/humble/Tutorials/Intermediate/Launch/Using-ROS2-Launch-For-Large-Projects.html) where we need to scroll down to the `return` part.
  - In contrast to [Nav 2](https://github.com/ros-planning/navigation2/blob/e2c8555ad352d69aa3da1e670b888d00f6eaeb3a/nav2_bringup/launch/navigation_launch.py) where each parameters are stored in a variable and added later. I don't think the naming of these parameters are necessary. Also, adding them later is error prone and some parameters may be easily overlooked.
- The `ld = LaunchDescription` allow easier debugging by adding/removing a node by modifying a single line (and without require scrolling to find the node). It also allows quickly view all nodes used in the launch file without needed to scroll a lot.
  - This style is slightly more error prone than the style in [ROS2 tutorial](https://docs.ros.org/en/humble/Tutorials/Intermediate/Launch/Using-ROS2-Launch-For-Large-Projects.html), since we need to name each node and remember to add them afterwards.
  - However, I think the benefits outweigh the drawbacks.

YuZhong-Chen commented on kobuki_ws/src/kobuki_gazebo/launch/gazebo.launch.py:
[Here](https://github.com/husky/husky/blob/729f8aa45ccd86fa33a05e07ef698c52c451cd9c/husky_gazebo/launch/gazebo.launch.py) is the reference, and I have added this link in that commit.

YuZhong-Chen replied:
`SetEnvironmentVariable` is required because Gazebo needs to know where to find the model.
You can also set that environment variable using the command: `export GAZEBO_MODEL_PATH=...` in termainl.
( I set the variable in the launch file since it is more convenient. )

Reference: https://classic.gazebosim.org/tutorials?tut=components&cat=get_started

j3soon replied:
Maybe change:

```
EnvironmentVariable("GAZEBO_MODEL_PATH", default_value=""),
LaunchConfiguration("GAZEBO_MODEL_PATH"),
```

to

```
EnvironmentVariable("GAZEBO_MODEL_PATH", default_value=""),
":",
LaunchConfiguration("GAZEBO_MODEL_PATH"),
```

to prevent a special case where both environment variable and the launch argument is set?

YuZhong-Chen replied:
Thank you for the reminder. 
The original version could lead to potential issues. I used the fix you provided !

YuZhong-Chen commented on kobuki_ws/src/kobuki_ros/kobuki_description/urdf/kobuki.urdf.xacro:
```diff
@@ -221,7 +221,9 @@
 
     <!-- Kobuki Gazebo simulation details -->
-    <xacro:kobuki_sim/>
+    <xacro:if value="$(arg is_sim)">
+        <xacro:kobuki_sim/>
+    </xacro:if>
```
It is fine to remove these three lines if we only get data from a real robot. However, if we are using Gazebo for simulation, these lines are required for Gazebo configuration.

Additionally, if we are not opening Gazebo when using a real robot, it is okay to remove the if condition. The if condition is used to prevent misuse.

j3soon replied:
Since I'm not familiar with Gazebo, is there a reference on this requirement? Thanks!

YuZhong-Chen replied:
I believe there isn't a comprehensive tutorial on setting up a robot in Gazebo with ROS 2. Many tutorials are written for ROS 1. Therefore, I frequently refer to the [ROS 1 documentation](https://classic.gazebosim.org/tutorials?tut=ros_gzplugins) and find corresponding [migration methods](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki) for ROS 2.

For example, when configuring the differential drive plugin, I refer to the [documentation](https://classic.gazebosim.org/tutorials?tut=ros_gzplugins#DifferentialDrive) to understand the relevant parameters. Afterward, I check the [migration wiki](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-Migration:-Diff-drive) for differential drive to adapt it to ROS 2. While this approach works for me, it can be a bit cumbersome.

YuZhong-Chen replied:
Fortunately, the original configuration has already been migrated to ROS 2, 
and I only need to fix certain parts, such as the TF hierarchy.

YuZhong-Chen replied:
I'm not sure if I fully answered the question, you might be specifically interested in the tag `xacro:kobuki_sim`. 
This tag is utilized to include the Gazebo configuration. 
It is located in `kobuki_gazebo.urdf.xacro`, in the same folder as `kobuki.urdf.xacro`.

j3soon replied:
Yes I'm interested in the `xacro:kobuki_sim` tag. I overlooked the use of it in `kobuki_gazebo.urdf.xacro`.

I also appreciate the additional documents and guidance on setting up the gazebo plugin!

YuZhong-Chen commented on kobuki_ws/docker/Dockerfile:
```diff
@@ -54,6 +54,12 @@ RUN apt-get update && apt-get install -y \
     ros-$ROS_DISTRO-ecl-tools \
     ros-$ROS_DISTRO-diagnostics \
+    #
+    # Install navigation dependencies
+    ros-$ROS_DISTRO-navigation2 \
+    ros-$ROS_DISTRO-nav2-bringup \
+    ros-$ROS_DISTRO-slam-toolbox \
+    ros-$ROS_DISTRO-robot-localization \
```
`robot-localization` is used to provide the tf of `map` -> `odom`. ( The EKF node )

j3soon replied:
Do you think we should add the 3 reference links in the comments?

YuZhong-Chen replied:
Sorry, I misunderstood the meaning and added these links to that commit message. 
I will include them in the comments in the Dockerfile.

YuZhong-Chen commented on kobuki_ws/src/kobuki_rviz/rviz/view_model.rviz:
I think there is no need to check the `.rviz` file since it is generated by RViz. Also, manually modifying the configuration in the file may not work sometimes. 

For instance, I had manually changed the value of `Marker Scale` to 0.3, opened RViz, and then attempted to save the configuration through RViz. It turned out that RViz saved the value back as 0.30000001192092896.

I believe there are two options here:

1. Restore every modification to the original settings. (It may be inconvenient when using the RViz.)
2. Ignore every difference in the `.rviz` file and consider it as added to `.gitignore`. (We still need the file, but we won't pay attention to its changes.)

j3soon replied:
Thank you! I'm convinced that there is no need to review the RVIZ file. We can regard it as some binary config file that isn't meant to be read.

I think it's good to keep it as is.

YuZhong-Chen commented on kobuki_ws/src/kobuki_rviz/rviz/gazebo.rviz:
I think the purpose of `.rviz` is not for reading, so it doesn't need to be readable in the diff.

---

Separating it into two commits may not provide any benefit, as the original file can be considered as directly generated by RViz. It is not copied from anywhere.

j3soon replied:
As mentioned in a previous commit. I'm convinced that we don't need to review the RVIZ file.

j3soon commented on kobuki_ws/docker/Dockerfile:
```diff
@@ -45,6 +45,17 @@ RUN apt-get update && apt-get install -y \
     ros-$ROS_DISTRO-rviz2 \
     && rm -rf /var/lib/apt/lists/*
 
+# Install Kobuki dependencies
```
I didn't see the references yet. Perhaps modifications were made but were somehow inadvertently removed?

YuZhong-Chen replied:
Same as the other modification, I have added the comment in that commit message. 
I will add them to the Dockerfile soon. Sorry for the misunderstanding.

j3soon commented on kobuki_ws/README.md:
```diff
@@ -0,0 +1,132 @@
+
+<div align="center">
+    <a href="./">
+        <img src="./figure/launch_robot_in_Gazebo.png" width="80%"/>
```
The screenshot is not up to date

j3soon reviewed (changes requested):
The current simulation code seems to work in most cases. However, it seems like the lidar data only contains scans in certain directions, and doesn't consider obstacles very near Kobuki. Please refer to the following video:

https://github.com/j3soon/ros2-essentials/assets/20457146/6691533a-8521-49ea-8f69-50c9b1edd12d

In addition I suggest the below changes by following #10:
- Simply mount `/dev` to prevent ttyUSB numbering issue
- Install udev rules inside the docker image
- Use built-in `TARGETARCH` env var for multi-arch builds
- Add keyboard teleop

For cross-building arm images on x86, refer to: https://www.stereolabs.com/docs/docker/building-arm-container-on-x86

YuZhong-Chen commented:
Hi, I have fixed all the issues mentioned above and added the new feature as well. I have tested this workspace on both my laptop and NV Orin, and it works perfectly on both platforms. Lastly, I want to express my gratitude for the information on cross-compilation; it is incredibly helpful and has saved me a lot of time during the package-building process. I have added the steps for cross-compilation at the end of the Readme. However, I'm unsure if I've explained each step clearly enough. I did my best to make it understandable. XD

j3soon commented:
As discussed in today's offline meeting, please update this PR to include the kobuki driver in the Docker image to minimize compilation time. This can be isolated in a separate workspace layer for ease of development. See: 9dc57e8f8aeedce726490721dd43e72ca7ad64b7

YuZhong-Chen commented on kobuki_ws/docker/kobuki_driver_ws/COLCON_IGNORE:
The default colcon compilation behavior is to build all packages found in the current directory. You can refer to the [documentation](https://colcon.readthedocs.io/en/released/reference/discovery-arguments.html) for more information. If we don't want to use the `COLCON_IGNORE` file, we need to add the flags `--paths` or `--base-paths` when using colcon build. I believe this may be a bit inconvenient, and the current solution is sufficient.

j3soon replied:
If I understand correctly, after mounting the project directory into docker and executing `colcon` build commands, we only want it to compile code in the `src` directory. However, the code in the `docker` directory will also be compiled. Therefore, the `COLCON_IGNORE` file is needed.

YuZhong-Chen replied:
You are right, this is why I placed the `COLCON_IGNORE` file inside the Kobuki driver workspace. 

In ROS1, `catkin` only compiles the code inside the `src` folder. However, in ROS2, `colcon` compiles all packages in the base directory by default. Therefore, the workspace in the Docker folder will be compiled unless there is a `COLCON_IGNORE` file.

j3soon reviewed (commented):
Thank you! The code and simulation examples look good to me. I'll find some time to test it on real hardware and then merge this.

j3soon reviewed (approved):
Current version can control real kobuki by either amd64 or arm64 system.
